proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2934,6 +2939,105 @@ lemma reply_remove_ccorres:
+
+lemma ctcb_relation_BlockedOnSend_blockingObject:
+  assumes "ctcb_relation tcb ctcb"
+  assumes "BlockedOnSend bo bib bicg bicgr biic = tcbState tcb"
+  shows "blockingObject_CL (thread_state_lift (tcbState_C ctcb)) = bo"
+  using assms
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma BlockedOnReceive_ep_at':
+  assumes "valid_objs' s"
+  assumes "st_tcb_at' ((=) (BlockedOnReceive bo bicg ro)) t s"
+  shows "ep_at' bo s"
+  using assms
+  apply -
+  apply (drule (1) tcb_in_valid_state')
+  by (fastforce simp: obj_at'_def valid_tcb_state'_def)
+
+lemma BlockedOnSend_ep_at':
+  assumes "valid_objs' s"
+  assumes "st_tcb_at' ((=) (BlockedOnSend bo bib bicg bicgr biic)) t s"
+  shows "ep_at' bo s"
+  using assms
+  apply -
+  apply (drule (1) tcb_in_valid_state')
+  by (fastforce simp: obj_at'_def valid_tcb_state'_def)
+
+lemma sym_ref_BlockedOnReceive_replyObject_linked:
+  assumes sy: "sym_refs (state_refs_of' s)"
+  assumes vo: "valid_objs' s"
+  assumes st: "st_tcb_at' ((=) (Structures_H.thread_state.BlockedOnReceive bo bicg (Some ro))) thread s"
+  shows "obj_at' (\<lambda>reply. replyTCB reply = Some thread) ro s"
+proof -
+  from vo and st have ra:
+    "reply_at' ro s"
+    by - (drule (1) tcb_in_valid_state', fastforce simp: obj_at'_def valid_tcb_state'_def)
+  from st have ko:
+    "\<exists>tcb. ko_at' tcb thread s \<and> tcbState tcb = BlockedOnReceive bo bicg (Some ro)"
+    by (clarsimp simp: st_tcb_at'_def obj_at'_def)
+  from sy and vo and st and ra and ko show ?thesis
+    apply -
+    apply (clarsimp simp: st_tcb_at'_def)
+    apply (frule (1) sym_ref_Receive_or_Reply_replyTCB')
+     apply simp
+    apply (clarsimp simp: obj_at'_def)
+    done
+qed
+
+lemma cancelIPC_threadSet_wp:
+  shows "threadSet (tcbFault_update (\<lambda>_. None)) t
+         \<lbrace>\<lambda>s. st_tcb_at' ((=) ts) t s \<and> invs' s
+              \<and> weak_sch_act_wf (ksSchedulerAction s) s \<and> sym_refs_asrt s\<rbrace>"
+proof -
+  have sy: "threadSet (tcbFault_update (\<lambda>_. None)) t \<lbrace>sym_refs_asrt\<rbrace>"
+    by - (unfold sym_refs_asrt_def, wpsimp wp: threadSet_state_refs_of', simp,
+          (force simp: tcb_bound_refs'_def)+)
+  show ?thesis
+    by (wpsimp wp: threadSet_fault_invs' threadSet_pred_tcb_no_state
+                   weak_sch_act_wf_lift tcb_in_cur_domain'_lift sy)
+qed
+
+lemma thread_state_to_tsType_eq_BlockedOnReceive:
+  "(thread_state_to_tsType ts = scast ThreadState_BlockedOnReceive)
+      = (\<exists>bo bicg ro. ts = BlockedOnReceive bo bicg ro)"
+  by (cases ts, simp_all add: ThreadState_defs)
+
+lemma thread_state_to_tsType_eq_BlockedOnSend:
+  "(thread_state_to_tsType ts = scast ThreadState_BlockedOnSend)
+      = (\<exists>bo bib bicg bicgr biic. ts = BlockedOnSend bo bib bicg bicgr biic)"
+  by (cases ts, simp_all add: ThreadState_defs)
+
 lemma cancelIPC_ccorres1:
```

Comment:
I have more comments for the proof of this lemma, but it would be useful to know how much of the proof is copied, as a bunch of them will be style-and-compression related.

Reply:
Some blocks, along with the overall structure, are copied from the non-MCS proof, but I've touched most of it. Sort of a ship of Theseus-like situation.